module/zfs/zfs_ioctl.c

```diff
@@ -2342,26 +2358,44 @@ zfs_ioc_snapshot_list_next(zfs_cmd_t *zc)
 		return (SET_ERROR(ESRCH));
 	}
 
-	error = dmu_snapshot_list_next(os,
-	    sizeof (zc->zc_name) - strlen(zc->zc_name),
-	    zc->zc_name + strlen(zc->zc_name), &zc->zc_obj, &zc->zc_cookie,
-	    NULL);
+	while (error == 0) {
+		error = dmu_snapshot_list_next(os,
+		    sizeof (zc->zc_name) - strlen(zc->zc_name),
+		    zc->zc_name + strlen(zc->zc_name), &zc->zc_obj,
+		    &zc->zc_cookie, NULL);
+		if (error == ENOENT) {
+			error = SET_ERROR(ESRCH);
+			break;
+		} else if (error != 0) {
+			break;
+		}
 
-	if (error == 0 && !zc->zc_simple) {
-		dsl_dataset_t *ds;
-		dsl_pool_t *dp = os->os_dsl_dataset->ds_dir->dd_pool;
+		if (zc->zc_simple)
+			break;
 
-		error = dsl_dataset_hold_obj(dp, zc->zc_obj, FTAG, &ds);
-		if (error == 0) {
-			objset_t *ossnap;
+		if ((error = dsl_dataset_hold_obj(dmu_objset_pool(os),
+		    zc->zc_obj, FTAG, &ds)) != 0)
+			break;
 
-			error = dmu_objset_from_ds(ds, &ossnap);
-			if (error == 0)
-				error = zfs_ioc_objset_stats_impl(zc, ossnap);
+		if ((skip_after != 0 && dsl_get_creationtxg(ds) > skip_after) ||
+		    (skip_to != 0 && dsl_get_creationtxg(ds) < skip_to)) {
+				dsl_dataset_rele(ds, FTAG);
+				/* undo snapshot name append */
+				*(strchr(zc->zc_name, '@') + 1) = '\0';
+				/* skip snapshot */
+				continue;
+		}
+
+		if ((error = dmu_objset_from_ds(ds, &ossnap)) != 0) {
 			dsl_dataset_rele(ds, FTAG);
+			break;
 		}
-	} else if (error == ENOENT) {
-		error = SET_ERROR(ESRCH);
+		if ((error = zfs_ioc_objset_stats_impl(zc, ossnap)) != 0) {
+			dsl_dataset_rele(ds, FTAG);
+			break;
+		}
+		dsl_dataset_rele(ds, FTAG);
+		break;
```

Comment:
all of these cases end with break except 1. For these I prefer to use goto in stead of a while loops, although I know they are a bad word for many people. Probably don't need to go through that though.